### ISSUE TITLE
[NativeAOT] App Store Connect warning ITMS-90999: Invalid executable 

### ISSUE THREAD

**Issue by rolfbjarne** [dotnet/runtime#94655]
_From @tipa on Thu, 02 Nov 2023 18:59:22 GMT_

I read about the NativeAOT experimental feature [here](https://devblogs.microsoft.com/dotnet/dotnet-8-performance-improvements-in-dotnet-maui/#nativeaot-on-ios) and wanted to give it a try right away.
The app can be downloaded via TestFlight and is significantly smaller than before (with the Mono AOT) - **awesome**! I also don't see any functional issues

### Steps to Reproduce

1. Add `<PublishAot>true</PublishAot>` to .csproj
2. `dotnet publish`
3. Upload resulting .ipa to App Store Connect

### Expected Behavior
No warning/error

### Actual Behavior
Email from Apple:
```
We identified one or more issues with a recent delivery for your app, "MyApp" 1.0.0 (1.0.0). Your delivery was successful, but you may wish to correct the following issues in your next delivery: 
ITMS-90999: Invalid executable - The executable in “MyApp.iOS.app” contains multiple segments with the executable permission bit set when it should only contain one. 
After you’ve corrected the issues, you can upload a new binary to App Store Connect. 
```
### Environment

.NET8 RC2.1

_Copied from original issue xamarin/xamarin-macios#19387_

**Comment by rolfbjarne** [comment]
_From @rolfbjarne on Mon, 06 Nov 2023 09:46:15 GMT_

I've never seen that ITMS message before, and according to Google not many other people has either (https://developer.apple.com/forums/thread/740085 is the only other hit).

Would you be able to attach/provide the .ipa you uploaded?

**Comment by rolfbjarne** [comment]
_From @tipa on Mon, 06 Nov 2023 10:41:02 GMT_

I sent you the .ipa file in a private message. Has Store submission of a NativeAot .ipa been tested by Microsoft?

**Comment by rolfbjarne** [comment]
_From @rolfbjarne on Mon, 06 Nov 2023 10:50:23 GMT_

> I sent you the .ipa file in a private message.

Thanks!

> Has Store submission of a NativeAot .ipa been tested by Microsoft?

Yes, but we might have missed warnings.

**Comment by rolfbjarne** [comment]
_From @rolfbjarne on Mon, 06 Nov 2023 10:55:12 GMT_

I wonder if this is the reason:

```
Load command 4
      cmd LC_SEGMENT_64
  cmdsize 152
  segname __THUNKS
   vmaddr 0x0000000100b84000
   vmsize 0x0000000000008000
  fileoff 10977280
 filesize 32768
  maxprot 0x00000005
 initprot 0x00000005
   nsects 1
    flags 0x0
Section
  sectname __thunks
   segname __THUNKS
      addr 0x0000000100b84000
      size 0x0000000000008000
    offset 10977280
     align 2^14 (16384)
    reloff 0
    nreloc 0
     flags 0x80000400
 reserved1 0
 reserved2 0
```

This is an executable segment, but the `__TEXT` segment is also executable.

@filipnavara any reason we can't put the `__thunks` section in the `__TEXT` segment?

**Comment by rolfbjarne** [comment]
_From @filipnavara on Mon, 06 Nov 2023 12:20:13 GMT_

> @filipnavara any reason we can't put the `__thunks` section in the `__TEXT` segment?

The reason is that the `__THUNKS` (TEXT) and `__THUNKS_DATA` (DATA) have to be next to each other. The sections are mapped multiple times to support marshalled delegates. The executable code references the data by relative offset and the data part is modified at runtime.

Fixed number of the marshalled delegates fits into the pre-generated code/data table, and then it gets mapped again at a different memory location. The second mapping needs to a) use the same relative offsets between the code and the data, b) still needs the code part to have valid code signature.

If Apple started blocking that we have a problem. I don't know on any method to ensure an executable layout that would allow the remapping.

**Comment by rolfbjarne** [comment]
_From @rolfbjarne on Mon, 06 Nov 2023 12:25:15 GMT_

> I don't know on any method to ensure an executable layout that would allow the remapping.

Isn't this what Mono has been doing for years (with no app store warnings)?

**Comment by rolfbjarne** [comment]
_From @filipnavara on Mon, 06 Nov 2023 12:35:05 GMT_

> Isn't this what Mono has been doing for years (with no app store warnings)?

I frankly have no idea. I think Mono may have some fixed limit... @lambdageek, any pointers how marshalled delegates are implemented on Mono/iOS?

**Comment by rolfbjarne** [comment]
_From @charlesroddie on Thu, 09 Nov 2023 18:41:11 GMT_

Also got this on uploading to testflight. Something new between dotnet8-rc1 and latest dotnet8.

**Comment by rolfbjarne** [comment]
_From @filipnavara on Sun, 12 Nov 2023 16:25:01 GMT_

The `vm_remap` trick in Mono is actually quite cool. I think it can be improved with the `VM_FLAGS_OVERWRITE` flag to avoid the retry loop and it may be quite reliable way to map the trampoline pages without having them as consecutive blocks in the executable file.

/cc @ivanpovazan

**Comment by jkotas** [comment]
Related to #82317 . cc @filipnavara

**Comment by filipnavara** [comment]
> Related to #82317 . cc @filipnavara

We already discussed how to fix it. Ivan or I will probably take a stab at it.

**Comment by neon-sunset** [comment]
Is there a chance this could get backported to 8.0.xxx if a fix becomes available?

**Comment by ivanpovazan** [comment]
> Is there a chance if a fix becomes available, it gets backported to 8.0.xxx?

We will consider it once we have the fix in place and will update the issue accordingly. Thank you for understanding.
